### PR TITLE
Fix and cleanup dead indirections

### LINKED ISSUE #36237: Runtime crash during JIT register allocation in .NET 5
Running the following program on .NET 5 (and master) with optimizations enabled crashes the runtime during register allocation:
```csharp
struct S0
{
    public bool F7;
}

class C1
{
    public S0 F7;
}

public class Program
{
    static C1 s_8;
    public static void Main()
    {
        uint[] vr0 = default(uint[]);
        if (s_8.F7.F7)
        {
            vr0 = vr0;
        }
    }
}
```

In checked builds the following assertion fires:
```
Assert failure(PID 18840 [0x00004998], Thread: 18212 [0x4724]): Assertion failed 'indirTree->TypeGet() != TYP_STRUCT' in 'Program:Main()' during 'Linear scan register alloc' (IL size 22)

    File: D:\dev\dotnet\runtime\src\coreclr\src\jit\lsraxarch.cpp Line: 2763
    Image: D:\dev\dotnet\runtime\artifacts\tests\coreclr\Windows_NT.x64.Checked\Tests\Core_Root\CoreRun.exe
```

cc @CarolEidt

### PR DESCRIPTION
**Pull request by CarolEidt**
The backend doesn't support a dead `TYP_STRUCT` indirection. Also, if a target register is needed, a `GT_IND` is better.

Fix #36237

### CODE REVIEW

**Comment by CarolEidt** [comment]
@dotnet/jit-contrib PTAL

**Comment by CarolEidt** [comment]
This results in mostly positive diffs:
| Arch  | OS      | What               | Size  | PerfScore | Methods Improved | Methods Regressed |
| ----  | :--     | :--                | --:   | --:       | --:              | --:               |
| Arm   | Windows | xgen fx+benchmarks | -96   | -165.60   | 19           | 0            |
| Arm64 | Windows | xgen fx+benchmarks | -16   | -41.60    | 2          | 0            |
| x64   | windows | xgen fx+benchmarks | -631  | -63.10    | 431          | 0            |
| x64   | Linux   | xgen fx+benchmarks | -2002 | -200.20   | 450          | 0            |
| x86   | windows | xgen fx+benchmarks | -75   | 278.09    | 58          | 9            |

Oddly, the problem in the regressed methods on x86 is a register pressure issue where previously we were spilling a variable in order to get an internal temp, and now we don't spill it, which results in worse code downstream because register pressure is higher.

File: src/coreclr/src/jit/codegenarmarch.cpp [hunk 1]
```diff
@@ -1500,19 +1500,19 @@ void CodeGen::genCodeForPhysReg(GenTreePhysReg* tree)
 // Return value:
 //    None
 //
-void CodeGen::genCodeForNullCheck(GenTreeOp* tree)
+void CodeGen::genCodeForNullCheck(GenTreeIndir* tree)
 {
+#ifdef TARGET_ARM
+    assert("GT_NULLCHECK isn't supported for Arm32; use GT_IND.");
```
Comment (sandreenko) [inline src/coreclr/src/jit/codegenarmarch.cpp:1506]:
```suggestion
    assert(!"GT_NULLCHECK isn't supported for Arm32; use GT_IND.");
```

**Review by sandreenko — APPROVED** [review]
LGTM, thank you for the cleanup.

**Comment by CarolEidt** [comment]
The 'Performance Linux x64 release coreclr JIT net5.0' leg has failed, but passed on the previous commit, with the only intervening change being arm32-only.

**Pull request merged**